test/hotspot/jtreg/compiler/itergvn/TestSplitPhiThroughMergeMem.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright (c) 2025, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/**
+ * @test
+ * @bug 8351833
+ * @summary Splitting Phi memory nodes through MergeMem nodes in PhiNode::Ideal
+ *          could sometimes result in a too large number of added nodes within
+ *          a single iteration of the main loop in PhaseIterGVN::optimize. This
+ *          test, reduced from TestScalarReplacementMaxLiveNodes, triggers an
+ *          assert added as part of the fix for the linked bug/issue (the
+ *          assert naturally triggers only before the fix). The test's ability
+ *          to trigger the issue is quite sensitive to the specific String
+ *          constants used. The current set of chosen String constants happened
+ *          to work particularly well.
+ * @run main/othervm -Xbatch
+ *                   -XX:CompileCommand=CompileOnly,compiler.itergvn.TestSplitPhiThroughMergeMem::test
+ *                   compiler.itergvn.TestSplitPhiThroughMergeMem
+ * @run main compiler.itergvn.TestSplitPhiThroughMergeMem
+ */
+
+package compiler.itergvn;
+
+public class TestSplitPhiThroughMergeMem {
+
+    public static void main(String[] args) {
+        for (int i = 0; i < 10_000; i++) {
+            int val = i % 50;
+            test(val == 0, val % 10, val % 20);
+        }
+    }
+
+    static void test(boolean flag, int param1, int param2) {
+        if (flag) {
+            new String("tenth" + param1);
+            new String("eleventh" + param2);
+            new String("fifteenth" + param2);
+            new String("sixteenth" + param1);
+            new String("seventeenth" + param1);
+            new String("nineteenth" + param2);
+            new String("tweenth" + param1);
+            new String("nineth" + param1);
+            new String("nineth" + param1);
+            new String("eighteenth" + param1);
+            new String("abcdef" + param2);
+            new String("ghijklmn" + param1);
+            new String("ghijklmn" + param1);
```

Comment:
This test illustrates an interesting behavior: C2 generates around 12 Kb of code for this rather infrequent code path (and the frequency can be further reduced without affecting C2's outcome). This seems to contradict C2's general philosophy of focusing the compilation effort (and code cache usage) on hot code. It would be interesting to investigate whether there is an opportunity to make some heuristic more execution-frequency aware here.

Reply:
Yes, for sure interesting. Let us create a separate RFE to investigate.

Reply:
I had a closer look at the example and found that the generation of large amounts of cold code is due to two factors:
1. (excessive?) use of `@ForceInline` in [StringConcatHelper](https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/lang/StringConcatHelper.java) and 
2.  generation of fast/path idioms for constructs within the cold inlined code (allocations in particular). 

For 1. we should evaluate whether forced inlining in  `StringConcatHelper` is warranted (in terms of cost/benefits). For 2., I filed [JDK-8354509](https://bugs.openjdk.org/browse/JDK-8354509).

Reply:
Thanks for filing [JDK-8354509](https://bugs.openjdk.org/browse/JDK-8354509). Should we file an RFE for 1. as well?